src/quera_ahs_utils/plotting.py

```diff
@@ -248,3 +244,182 @@ def plot_avg_density(densities, register, with_labels = True, custom_axes = None
     else:
         return None,ax
 
+
+
+#%%
+def visualize_UDG(positions,
+                  radii,
+                  ax=None,
+                  label_nodes=False,
+                  draw_disks=False,
+                  color_nodes=3,
+                  color_edges=0,
+                  scale=1,
+                  vertex_edge = False):
+    '''
+    Draws a unit disk graph.
+    positions - 2d positions of each vertex
+    radii     - Float. radius of each vertex. Vertices are connected if the unit disks overlap.
+                !! This is a factor of 2 from usual !!
+    draw_disks - If unit disks should be drawn.
+            True  : Draw disks for all vertices
+            False : Do not draw disks
+            list  : list of True False for each vertex
+    color_nodes - The color for each vertex.
+            int   : color assignment for all vertices.
+                0   > Black.            Indicates an inert / ground state Rydberg atom
+                1   > "Red"  #C2477F.   Indicates a Rydberg independent set excited atom
+                2   > "Pink" #FFE5E5    Indicates an inert / ground state Rydberg atom against a dark background
+                3   > "Blue" #6437FF    Default; used when a vertex is stateless
+            str > a color descriptor for every vertex. eg #8ad679
+            list  : color assignment for each vertex.
+                     Each element may be an int or a string.
+    
+    ax          - axis to plot on. If None, generates a new axis that scales
+                    so that every edge is a consistent size.
+    color_edges - The color for each edge.
+            int  : A style label for every edge
+                0   > Thick black line.   Used for geometric connections.
+                1   > Thin grey line.     Used for nongeometric connections.
+                2   > Thick pink line.    Used against dark backgrounds
+               -1   > No line.
+            tuple > (str, int). str indicates line color, int indicates line weight
+            dict : keys are edge labels, and values are style labels (int) or tuples.
+    label_nodes - label each vertex.
+    scale       - scaling of vertex size. Smaller number means more area
+    vertex_edge - Draw an edge around each vertex.
+    '''
+    
+    positions = np.array(positions)
+    N = len(positions)
+    
+    
+    
+    
+    
+    
+    # Construct the unit disk graph
+    G = nx.Graph((np.sqrt((positions[:,0:1] - positions[:,0:1].T)**2 + (positions[:,1::] - positions[:,1::].T)**2) + np.diag(np.ones(N))*1e99)<(2*radii))
+    # Relabel vertices
+    relabeling = {i:tuple(positions[i]) for i in range(N)}
+    G = nx.relabel.relabel_nodes(G, relabeling)
+    
+    pos = {tuple(a):a for a in positions}
+    
+    if hasattr(color_nodes,'__len__'):
```

Comment:
```suggestion
    if type(color_nodes) == list:
```
Considering the docstring explicitly says color_nodes can be of type list, it would be best to restrict type checking to it. Furthermore, I don't think just checking for the __len__ attribute is sufficient to ensure that indexing on the object is supported

Reply:
So color_nodes can be any kind of iterable? Then it should use `hasattr(color_nodes, `__iter__`). 

I proposed type(...) == list because the docstring had type "list" for the argument which I assumed meant the default python implementation. If you want to restrict it to numpy arrays you could do `type(...) == np.ndarray`.